### DISCUSSION TITLE
Can set the "encrypt filenames" of 7z?

### DISCUSSION THREAD

**Discussion by Polar-Pumpkin** [rikyoz/bit7z#237] [Q&A] [answered]
I created a password-protected archive, but I found that I just needed a password for extraction, and I was still free to view the list of files. So I searched for related methods and found that 7zip supports "encrypt file names", which requires a password when opening the archive, can I do this in bit7z?

**Comment by Polar-Pumpkin** [comment]
I found the `setFormatProperty` method, and based on the search, I found the `-mhe` command line parameter can encrypt the file header, thereby achieving the effect of encrypting the file name. At first I thought I needed to pass in `mhe`, and combined with the `tm` parameter mentioned in the document, I realized that I didn't need to specify `m`

**Reply by rikyoz** [reply, marked as answer]
Hi.
In addition to the `setFormatProperty` method, you might want to look at the [`setPassword`](https://github.com/rikyoz/bit7z/wiki/BitArchiveWriter#void-setpassword-const-tstring-password-bool-cryptheaders-) method overload, which takes a second boolean parameter specifying whether or not to encrypt the headers.
It's perhaps a cleaner way of enabling header encryption, as you don't need any knowledge of 7-Zip format properties to use it (but functionally it's equivalent to what you already did).

**Reply by Polar-Pumpkin** [reply]
Thank you so much for telling me this method! I thought this was a parameter for the encryption method, such as the encryption algorithm or something. Now I know that I misunderstood it 🤣

**Reply by rikyoz** [reply]
You're welcome!

> I thought this was a parameter for the encryption method, such as the encryption algorithm or something. Now I know that I misunderstood it 🤣

To be honest, I don't really like this method signature myself.
I have long since come to the conclusion that `bool` parameters should be avoided (at least most of the time).
I have half a mind to add a more explicit method overload using an enumeration parameter instead of a boolean, e.g. using an `enum` like this:
```cpp
enum struct EncryptionLevel {
    DataOnly, // Or ContentsOnly
    WholeArchive // Or just HeaderEncryption
};
//...
writer.setPassword("password", EncryptionLevel::WholeArchive);
```

I think this would be more readable than the current `writer.setPassword("password", true);`.
But I'm still not sure about the names of the enumeration values.
As they say, naming things is one of the hardest problems in computer science. 😅
